proposals/data-objects.md

```diff
@@ -0,0 +1,84 @@
+# Data objects
+
+* **Type**: Design proposal
+* **Authors**: Alexander Udalov, Roman Elizarov, Pavel Mikhailovskii
+* **Status**: Experimental since 1.7.20, stable since 1.8.0
+* **Discussion and feedback**: [#317](https://github.com/Kotlin/KEEP/issues/317)
+
+
+## Introduction
+
+When using algebraic data types, it’s annoying to declare singleton values because there’s no way to get automatic toString similar to data classes. E.g. on JVM:
+```kotlin
+sealed class MyList<T> {
+    data class Cons<T>(val value: T) : MyList<T>()
+    object Nil : MyList<Nothing>()
+}
+
+fun main() {
+    println(MyList.Cons(42))   // Cons(value=42)
+    println(MyList.Nil)        // test.MyList$Nil@1d251891
+}
+```
+
+A workaround for the user is either to declare `toString `manually, which leads to boilerplate, or use data class with a placeholder default argument, which seems bizarre.
+
+## Other aspects
+
+* Default toString implementation differs on JVM, JS, Native
+    * In JS, toString returns [object Object]
+    * Native behaves similarly to JVM, but inner classes are delimited with ., i.e. test.List*.*Nil@1d251891 in the example above
+* The @hash part is meaningless even for non-ADT objects, because object is a singleton
+* Can’t declare an empty @JvmRecord class in Kotlin (KT-48155 (https://youtrack.jetbrains.com/issue/KT-48155/JvmRecord-does-not-allow-one-to-define-no-parameters))
+    * Probably not a problem
```

Comment:
Let's add that Kotlin reflect `isData` method will now return `true` for `data object`? 

It might be useful for people to understand what consequences does annotating `object` with `data` bring